flytekit/types/flyte_directory.py

```diff
@@ -0,0 +1,212 @@
+
+import os
+import typing
+from pathlib import Path
+
+from flytekit.annotated.context_manager import FlyteContext
+from flytekit.annotated.type_engine import TypeEngine, TypeTransformer
+from flytekit.models import types as _type_models
+from flytekit.models.core import types as _core_types
+from flytekit.models.literals import Blob, BlobMetadata, Literal, Scalar
+from flytekit.models.types import LiteralType
+
+
+class FlyteDirectory(os.PathLike):
+    """
+    Please first read through the comments on the FlyteFile class as the implementation here is similar.
+
+    One thing to note is that the os.PathLike type that comes with Python was used as a stand-in for FlyteFile.
+    That is, if a task returns an os.PathLike, Flyte takes that to mean FlyteFile. There is no easy way to
+    distinguish an os.PathLike where the user means a File and where the user means a Directory. As such, if you
+    want to use a directory, you must declare all types as FlyteDirectory. You'll still be able to return a string
+    literal though instead of a full-fledged FlyteDirectory object assuming the str is a directory.
+
+    Use cases as inputs:
+        def t1(in1: FlyteDirectory):
+            ...
+
+        def t1(in1: FlyteDirectory["svg"]):
+            ...
+
+    As outputs:
+
+    The contents of this local directory will be uploaded to the Flyte store.
+        return FlyteDirectory("/path/to/dir/")
+
+        return FlyteDirectory["svg"]("/path/to/dir/", remote_path="s3://special/output/location")
+
+    Similar to the FlyteFile example, if you give an already remote location, it will not be copied to Flyte's
+    durable store, the uri will just be stored as is.
+
+        return FlyteDirectory("s3://some/other/folder")
+
+    Note if you write a path starting with http/s, if anything ever tries to read it (i.e. use the literal
+    as an input, it'll fail because the http proxy doesn't know how to download whole directories.
+
+    The format [] bit is still there because in Flyte, directories are stored as Blob Types also, just like files, and
+    the Blob type has the format field. The difference in the type field is represented in the ``dimensionality``
+    field in the ``BlobType``.
+    """
+
+    def __init__(self, path: str, downloader: typing.Callable = None, remote_directory=None):
+        """
+        :param path: The source path that users are expected to call open() on
+        :param downloader: Optional function that can be passed that used to delay downloading of the actual fil
+            until a user actually calls open().
+        :param remote_directory: If the user wants to return something and also specify where it should be uploaded to.
+        """
+
+        def noop():
+            ...
+
+        self._path = path
+        self._downloader = downloader or noop
+        self._downloaded = False
+        self._remote_directory = remote_directory
+        self._remote_source = None
+
+    def __fspath__(self):
```

Comment:
when is this invoked? For a directory, should we not support list-dir?
And should we download all at once, or allow a streaming download so that you can read as some are downloaded?

Reply:
I see it is invoked for listdir as well. I like that too.
So lets keep this simple, I agree. Something that downloads at once a very simple interface. We can later build a streamingDirectory?

Reply:
added comment about list dir.  Yeah a streaming directory of some kind in the future sounds good.  I think we'll have a better idea of how to implement things when the streaming story is clearer.